File: src/main/java/lekkit/scev/machine/storage/BuildrootDiskTemplate.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package lekkit.scev.machine.storage;

import net.minecraft.network.chat.Component;

/**
 * First-party preloaded-disk template: a minimal ext2 Linux rootfs
 * ({@code linux_rootfs.ext2}, 16 MiB, deterministic UUID
 * {@code deadbeef-cafe-babe-feed-facefacefeed}, label {@code SCEV_ROOTFS}).
 *
 * <p>Carries a Unix directory skeleton ({@code /etc}, {@code /bin},
 * {@code /sbin}, {@code /proc}, {@code /sys}, {@code /dev}, {@code /root},
 * {@code /tmp}, {@code /mnt}, {@code /boot}) plus:
 *
 * <ul>
 *   <li>{@code /README} — usage notes.</li>
 *   <li>{@code /etc/hostname} — {@code "scev-linux"}.</li>
 *   <li>{@code /etc/fstab} — proc/sysfs/devtmpfs/tmpfs mounts.</li>
 *   <li>{@code /etc/os-release} — distro metadata.</li>
 * </ul>
 *
 * <p>Once a workstation has a {@link lekkit.scev.items.PreloadedNvmeItem}
 * installed, the guest Linux kernel can mount it with:
 *
 * <pre>
 *     mount -t ext2 /dev/nvme0n1 /mnt
 * </pre>
 *
 * <p>The filesystem is validated with {@code e2fsck -f -n} at build time
 * and is fully standards-compliant — a real ext2, not a synthetic byte
 * pattern.
 *
 * <p><b>Expanding the rootfs with a real BusyBox userland</b> is a
 * forward-compatible change: replace the {@code linux_rootfs.ext2} asset
 * with a larger Buildroot-produced filesystem (bumping {@link #SIZE_MB}
 * if it grows). No code change needed.
 */
public final class BuildrootDiskTemplate implements ScevDiskTemplate {
    public static final BuildrootDiskTemplate INSTANCE = new BuildrootDiskTemplate();

    /** Classpath asset name under {@code /assets/scev/firmware/}. */
    public static final String ASSET_NAME = "linux_rootfs.ext2";

    /**
     * Declared disk size (MiB). Must be ≥ the actual asset size (16 MiB
     * today). The extra capacity is zero-initialized — ext2 is happy with
     * a sparse tail, though a future feature (auto-resize on first boot)
     * would expand the filesystem to fill the declared capacity.
     */
    public static final long SIZE_MB = 2048;

    /** Deterministic ext2 filesystem UUID set at asset build time. */
    public static final String FILESYSTEM_UUID = "deadbeef-cafe-babe-feed-facefacefeed";

    /** Deterministic ext2 volume label set at asset build time. */
    public static final String FILESYSTEM_LABEL = "SCEV_ROOTFS";

    private BuildrootDiskTemplate() {}

    @Override public String assetName() { return ASSET_NAME; }
    @Override public long sizeMb() { return SIZE_MB; }
    @Override public Component displayName() { return Component.literal("Buildroot Linux"); }
}

File: src/main/kotlin/lekkit/scev/machine/storage/BuildrootDiskTemplate.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package lekkit.scev.machine.storage

import net.minecraft.network.chat.Component

/**
 * First-party preloaded-disk template: a minimal ext2 Linux rootfs
 * (`linux_rootfs.ext2`, 16 MiB, deterministic label [FILESYSTEM_LABEL]
 * and UUID [FILESYSTEM_UUID]).
 *
 * Carries a Unix directory skeleton (`/etc`, `/bin`, `/sbin`, `/proc`,
 * `/sys`, `/dev`, `/root`, `/tmp`, `/mnt`, `/boot`) plus `/README`,
 * `/etc/hostname` (`"scev-linux"`), `/etc/fstab`, and `/etc/os-release`.
 *
 * Once installed on a [lekkit.scev.items.PreloadedNvmeItem], the guest
 * kernel can mount it with:
 *
 * ```
 * mount -t ext2 /dev/nvme0n1 /mnt
 * ```
 *
 * The filesystem is validated with `e2fsck -f -n` at build time. The
 * declared [SIZE_MB] is 2048 — must be ≥ the actual 16 MiB asset; the
 * extra capacity is zero-initialized, which ext2 handles as a sparse tail.
 * Replacing `linux_rootfs.ext2` with a larger Buildroot-produced filesystem
 * needs only [SIZE_MB] bumped to match; no other code change.
 */
object BuildrootDiskTemplate : ScevDiskTemplate {
    const val ASSET_NAME = "linux_rootfs.ext2"

    /**
     * Declared capacity in MiB. The shipped `linux_rootfs.ext2` is much
     * smaller (64 MiB BusyBox rootfs from `tools/buildroot/build-kernel.sh`,
     * or a 16 MiB skeleton on older builds). [lekkit.scev.server.StorageManager.initImage]
     * sparse-extends the per-UUID image to this value after the template
     * copy, so from the guest's perspective the NVMe block device is 1 GiB
     * — matching what the item tooltip advertises and what [lekkit.scev.items.NvmeItem.SIZE_MB]
     * declares for blank NVMes. The `/init` pivot script in
     * `tools/buildroot/rootfs_overlay/init` calls `resize2fs` on first boot
     * to grow the filesystem to fill that advertised capacity.
     */
    const val SIZE_MB = 1024L

    const val FILESYSTEM_UUID = "deadbeef-cafe-babe-feed-facefacefeed"
    const val FILESYSTEM_LABEL = "SCEV_ROOTFS"

    override fun assetName(): String = ASSET_NAME
    override fun sizeMb(): Long = SIZE_MB
    override fun displayName(): Component = Component.literal("Buildroot Linux")

    /**
     * Intended pairing is [lekkit.scev.machine.firmware.LinuxFirmware] plus
     * the initramfs pivot script in `tools/buildroot/rootfs_overlay/init`:
     * kernel loads from flash, initramfs switch_roots to this disk,
     * pid 1 lives on ext4. The declaration is the hook that makes the
     * parser emit `root=<rootDevice()> rw rootwait`.
     *
     * The asset is a **raw** ext filesystem (genext2fs output, no MBR
     * wrapping), mountable at the whole-disk device — so this template
     * keeps the interface default of `/dev/nvme0n1` for [rootDevice].
     * Contrast [AlpineDiskTemplate], whose build wraps ext4 in an MBR
     * partition table and therefore overrides rootDevice to p1.
     */
    override fun hasRootFilesystem(): Boolean = true

    /**
     * No extlinux.conf / on-disk kernel — U-Boot would find nothing to boot.
     * Pair with [lekkit.scev.machine.firmware.LinuxFirmware], not
     * [lekkit.scev.machine.firmware.OpenFirmware].
     */
    override fun isBootable(): Boolean = false
}
